File: app/src/main/java/fr/nihilus/music/ui/artists/ArtistDetailFragment.java
/*
 * Copyright 2017 Thibault Seisel
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.nihilus.music.ui.artists;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.provider.MediaStore.Audio.AlbumColumns;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v4.media.MediaBrowserCompat.MediaItem;
import android.support.v4.media.MediaBrowserCompat.SubscriptionCallback;
import android.support.v7.widget.GridLayoutManager;
import android.util.Log;
import android.view.View;

import java.util.List;

import dagger.android.support.AndroidSupportInjection;
import fr.nihilus.music.Constants;
import fr.nihilus.music.R;
import fr.nihilus.music.di.ActivityScoped;
import fr.nihilus.music.library.BrowserViewModel;
import fr.nihilus.music.ui.albums.AlbumDetailActivity;
import fr.nihilus.recyclerfragment.RecyclerFragment;

@ActivityScoped
public class ArtistDetailFragment extends RecyclerFragment
        implements ArtistDetailAdapter.OnMediaItemSelectedListener {

    public static final String BACKSTACK_ENTRY = "artist_detail";

    private static final String TAG = "ArtistDetailFragment";
    private static final String KEY_ARTIST = "artist";
    private static final String KEY_ARTIST_DETAIL = "artist_detail";

    private MediaItem mPickedArtist;
    private ArtistDetailAdapter mAdapter;

    private BrowserViewModel mViewModel;

    private final SubscriptionCallback mCallback = new SubscriptionCallback() {
        @Override
        public void onChildrenLoaded(@NonNull String parentId, @NonNull List<MediaItem> children) {
            Log.d(TAG, "onChildrenLoaded: loaded " + children.size() + " items");
            mAdapter.updateItems(children);
            setRecyclerShown(true);
        }
    };

    public static ArtistDetailFragment newInstance(MediaItem artist) {
        Bundle args = new Bundle(2);
        args.putInt(Constants.FRAGMENT_ID, R.id.action_artists);
        args.putParcelable(KEY_ARTIST, artist);
        ArtistDetailFragment fragment = new ArtistDetailFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onAttach(Context context) {
        AndroidSupportInjection.inject(this);
        super.onAttach(context);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Bundle args = getArguments();
        if (args == null || (mPickedArtist = args.getParcelable(KEY_ARTIST)) == null) {
            throw new IllegalStateException("Caller must specify the artist to display.");
        }

        mAdapter = new ArtistDetailAdapter(this);
        mAdapter.setOnMediaItemSelectedListener(this);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        final int spanCount = getResources().getInteger(R.integer.artist_grid_span_count);
        GridLayoutManager manager = new GridLayoutManager(getContext(), spanCount);
        getRecyclerView().setHasFixedSize(true);
        getRecyclerView().setLayoutManager(manager);
        manager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
            @Override
            public int getSpanSize(int position) {
                Bundle extras = mAdapter.get(position).getDescription().getExtras();
                if (extras != null) {
                    boolean isAlbum = extras.getString(AlbumColumns.ALBUM_KEY, null) != null;
                    return isAlbum ?  1 : spanCount;
                }
                return 1;
            }
        });
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        mViewModel = ViewModelProviders.of(getActivity()).get(BrowserViewModel.class);
        setAdapter(mAdapter);
    }

    @Override
    public void onStart() {
        super.onStart();
        getActivity().setTitle(mPickedArtist.getDescription().getTitle());
        mViewModel.subscribe(mPickedArtist.getMediaId(), mCallback);
    }

    @Override
    public void onStop() {
        mViewModel.unsubscribe(mPickedArtist.getMediaId());
        super.onStop();
    }

    @Override
    public void onAlbumSelected(ArtistDetailAdapter.AlbumHolder holder, MediaItem album) {
        ActivityOptionsCompat options = ActivityOptionsCompat.makeSceneTransitionAnimation(
                getActivity(), holder.albumArt, AlbumDetailActivity.ALBUM_ART_TRANSITION_NAME);
        Intent albumDetailIntent = new Intent(getContext(), AlbumDetailActivity.class);
        albumDetailIntent.putExtra(AlbumDetailActivity.ARG_PICKED_ALBUM, album);
        albumDetailIntent.putExtra(AlbumDetailActivity.ARG_PALETTE, holder.colors);
        startActivity(albumDetailIntent, options.toBundle());
    }

    @Override
    public void onTrackSelected(ArtistDetailAdapter.TrackHolder holder, MediaItem track) {
        String mediaId = track.getMediaId();
        if (mediaId == null) {
            throw new AssertionError("Track should have a mediaId");
        }

        mViewModel.playFromMediaId(track.getMediaId());
    }
}

File: app/src/main/java/fr/nihilus/music/ui/artists/ArtistDetailFragment.kt
/*
 * Copyright 2018 Thibault Seisel
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.nihilus.music.ui.artists

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.ActivityOptionsCompat
import android.support.v4.media.MediaBrowserCompat.MediaItem
import android.support.v7.widget.GridLayoutManager
import android.view.View
import dagger.android.support.AndroidSupportInjection
import fr.nihilus.music.R
import fr.nihilus.music.client.BrowserViewModel
import fr.nihilus.music.di.ActivityScoped
import fr.nihilus.music.media.Constants
import fr.nihilus.music.ui.BaseAdapter
import fr.nihilus.music.ui.albums.AlbumDetailActivity
import fr.nihilus.music.ui.albums.AlbumPalette
import fr.nihilus.music.ui.holder.AlbumHolder
import fr.nihilus.recyclerfragment.RecyclerFragment
import javax.inject.Inject

@ActivityScoped
class ArtistDetailFragment : RecyclerFragment(), BaseAdapter.OnItemSelectedListener {

    private lateinit var pickedArtist: MediaItem
    private lateinit var adapter: ArtistDetailAdapter

    private lateinit var viewModel: BrowserViewModel

    @Inject lateinit var defaultAlbumPalette: AlbumPalette

    override fun onAttach(context: Context?) {
        AndroidSupportInjection.inject(this)
        super.onAttach(context)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        pickedArtist = checkNotNull(arguments?.getParcelable(KEY_ARTIST)) {
            "Callers must specify the artist to display."
        }

        adapter = ArtistDetailAdapter(this, defaultAlbumPalette, this)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val spanCount = resources.getInteger(R.integer.artist_grid_span_count)
        val manager = GridLayoutManager(context, spanCount)
        recyclerView.setHasFixedSize(true)
        recyclerView.layoutManager = manager

        manager.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {

            override fun getSpanSize(position: Int): Int {
                val viewType = adapter.getItemViewType(position)
                return if (viewType == R.id.view_type_album) 1 else spanCount
            }
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        viewModel = ViewModelProviders.of(activity!!).get(BrowserViewModel::class.java)
        setAdapter(adapter)

        viewModel.subscribeTo(pickedArtist.mediaId!!).observe(this, Observer {
            adapter.submitList(it.orEmpty())
            setRecyclerShown(true)
        })
    }

    override fun onStart() {
        super.onStart()
        activity!!.title = pickedArtist.description.title
    }

    override fun onItemSelected(position: Int, actionId: Int) {
        val selectedItem = adapter.getItem(position)
        val holder = recyclerView.findViewHolderForAdapterPosition(position)
        when (holder.itemViewType) {
            R.id.view_type_track -> onTrackSelected(selectedItem)
            R.id.view_type_album -> onAlbumSelected(holder as AlbumHolder, selectedItem)
        }
    }

    private fun onAlbumSelected(holder: AlbumHolder, album: MediaItem) {
        val options = ActivityOptionsCompat.makeSceneTransitionAnimation(
            activity!!, holder.transitionView, AlbumDetailActivity.ALBUM_ART_TRANSITION_NAME
        )

        val albumDetailIntent = Intent(context, AlbumDetailActivity::class.java).apply {
            putExtra(AlbumDetailActivity.ARG_PICKED_ALBUM, album)
            putExtra(AlbumDetailActivity.ARG_PALETTE, holder.colorPalette)
        }

        startActivity(albumDetailIntent, options.toBundle())
    }

    private fun onTrackSelected(track: MediaItem) {
        val mediaId = track.mediaId ?: error("Track should have a mediaId")
        viewModel.post { it.transportControls.playFromMediaId(mediaId, null) }
    }

    companion object Factory {

        private const val KEY_ARTIST = "artist"

        fun newInstance(artist: MediaItem) = ArtistDetailFragment().apply {
            arguments = Bundle(2).apply {
                putInt(Constants.FRAGMENT_ID, R.id.action_artists)
                putParcelable(KEY_ARTIST, artist)
            }
        }
    }
}
